File: packages/frontend-web/src/lib/api/contract-fixing.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient } from '@/lib/api-client'

/**
 * Kontrakt-Fixierung & MATIF-Bewertung (DOM-CON-004.2).
 * Teilfixierung MATIF-bepreister Kontrakte + Mark-to-Market gegen Marktnotierung.
 * Backend: app/api/v1/endpoints/contract_fixing.py
 */

export type Fixing = {
  fixing_no: number
  menge: number | null
  matif_price: number | null
  praemie: number | null
  effektiv_preis: number | null
  datum: string | null
  referenz: string | null
  notiz: string | null
  bediener: string | null
}

export type FixingPosition = {
  position_no: number
  line_id: string
  artikel: string | null
  bezeichnung: string | null
  is_matif: boolean
  menge_kontrakt: number | null
  fixiert: number | null
  offen_zu_fixieren: number | null
  fixierungsgrad_pct: number
  avg_fixpreis_effektiv: number | null
  symbol: string | null
  notierung: { preis: number; datum: string; markt_effektiv: number; quelle: string | null } | null
  bewertung_fixiert_eur: number | null
  marktwert_offen_eur: number | null
  fixings: Fixing[]
}

export type FixingWorkspace = {
  found: boolean
  detail?: string
  contract_no?: string
  contract_type?: string
  party_id?: string | null
  einheit?: string | null
  status?: string | null
  pricing?: { modell: string | null; praemie_typ: string | null; praemie_wert: number | null; basis: string | null }
  positionen?: FixingPosition[]
  summary?: {
    menge_kontrakt: number
    fixiert: number
    offen_zu_fixieren: number
    fixierungsgrad_pct: number
    bewertbar: boolean
    bewertung_fixiert_eur: number | null
    marktwert_offen_eur: number | null
  }
}

export type CreateFixingInput = {
  kontrakt: string
  position?: string
  menge: number
  matifPreis: number
  praemie?: number
  matifReferenz?: string
  notiz?: string
}

const BASE = '/api/v1/contracts'

export function useFixingWorkspace(kontrakt: string | null) {
  return useQuery({
    queryKey: ['contract-fixing', 'workspace', kontrakt],
    enabled: !!kontrakt,
    queryFn: async () => {
      const res = await apiClient.get<FixingWorkspace>(`${BASE}/fixing/workspace`, { params: { kontrakt } })
      return res.data
    },
  })
}

export function useCreateFixing() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (input: CreateFixingInput) =>
      (await apiClient.post<{ ok: boolean; fixing_no: number; offen_nach: number }>(`${BASE}/fixing`, input)).data,
    onSuccess: (_d, vars) => {
      void qc.invalidateQueries({ queryKey: ['contract-fixing', 'workspace', vars.kontrakt] })
    },
  })
}

export function useUpsertQuote() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (input: { symbol: string; preis: number; datum?: string }) =>
      (await apiClient.post<{ ok: boolean; symbol: string; price: number }>(`${BASE}/matif-quote`, input)).data,
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: ['contract-fixing', 'workspace'] })
    },
  })
}

File: app/api/v1/endpoints/contract_fixing.py
"""Kontrakt-Fixierung & MATIF-Bewertung (DOM-CON-004.2).

Schreib-/Lese-Endpunkte für die Teilfixierung MATIF-bepreister Kontrakte und die
Mark-to-Market-Bewertung gegen die jüngste Marktnotierung.
"""

from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.tenant import get_tenant_id
from app.services.contract_fixing_service import ContractFixingService, FixingActionError

router = APIRouter(prefix="/contracts", tags=["contracts", "kontrakte", "agrar"])


@router.get("/fixing/workspace", summary="Fixierungs-Arbeitsraum je Kontrakt (fixiert/offen + Bewertung)")
def workspace(
    kontrakt: str = Query(..., description="Kontraktnummer oder -ID"),
    db: Session = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
) -> dict[str, Any]:
    return ContractFixingService(db, tenant_id).workspace(kontrakt)


@router.get("/fixing/list", summary="Fixierungen eines Kontrakts")
def list_fixings(
    kontrakt: str = Query(..., description="Kontraktnummer oder -ID"),
    db: Session = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
) -> dict[str, Any]:
    return {"items": ContractFixingService(db, tenant_id).list_fixings(kontrakt)}


class FixingIn(BaseModel):
    kontrakt: str
    position: Optional[str] = None          # Position-Nr oder line_id (bei mehreren Positionen Pflicht)
    menge: float
    matifPreis: float
    praemie: Optional[float] = None         # Default = Kontrakt-Prämie
    matifReferenz: Optional[str] = None
    notiz: Optional[str] = None
    bediener: Optional[str] = None


@router.post("/fixing", summary="Teilfixierung anlegen (Menge zu MATIF-Preis + Prämie)")
def create_fixing(
    body: FixingIn,
    db: Session = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
) -> dict[str, Any]:
    try:
        return ContractFixingService(db, tenant_id).create_fixing(
            contract_ref=body.kontrakt, line_ref=body.position, quantity=body.menge,
            matif_price=body.matifPreis, premium=body.praemie,
            matif_reference=body.matifReferenz, note=body.notiz, bediener=body.bediener or "KIM",
        )
    except FixingActionError as exc:
        raise HTTPException(status_code=422, detail=str(exc))


class QuoteIn(BaseModel):
    symbol: str
    preis: float
    datum: Optional[str] = None             # ISO-Datum; Default heute
    einheit: Optional[str] = None
    quelle: Optional[str] = None


@router.post("/matif-quote", summary="MATIF-Notierung erfassen/aktualisieren (Bewertungsbasis)")
def upsert_quote(
    body: QuoteIn,
    db: Session = Depends(get_db),
    tenant_id: str = Depends(get_tenant_id),
) -> dict[str, Any]:
    from datetime import date

    qd = None
    if body.datum:
        try:
            qd = date.fromisoformat(body.datum[:10])
        except ValueError:
            raise HTTPException(status_code=422, detail="Ungültiges Datum (erwartet ISO YYYY-MM-DD).")
    try:
        return ContractFixingService(db, tenant_id).upsert_quote(
            symbol=body.symbol, quote_date=qd, price=body.preis,
            unit=body.einheit, source=body.quelle or "manual",
        )
    except FixingActionError as exc:
        raise HTTPException(status_code=422, detail=str(exc))
